fpdf/html.py

```diff
@@ -236,28 +266,44 @@ def __init__(
         self.li_tag_indent = li_tag_indent
         self.dd_tag_indent = dd_tag_indent
         self.ul_bullet_char = ul_bullet_char
+        self.heading_sizes = dict(**DEFAULT_HEADING_SIZES)
+        if heading_sizes:
+            self.heading_sizes.update(heading_sizes)
+        self.pre_code_font = pre_code_font
+        self.warn_on_tags_not_matching = warn_on_tags_not_matching
+
+        # We operate in a local context and will only temporarily switch to the outer one for rendering.
+        # This is necessary because of the deferred execution of text region writes. Changing fonts and
+        # colors in here must not affect the output and/or interfere with the settings during rendering.
+        # The TOC gets rendered outside of our scope, so we need to set a font first, in order to
+        # ensure that the TOC has one available.
+        # If a font was defined previously, we reinstate that seperately after we're finished here.
+        # In this case the TOC will be rendered with that font and not ours. But adding a TOC tag only
+        # makes sense if the whole document gets converted from HTML, so this should be acceptable.
         self.style = dict(b=False, i=False, u=False)
-        self.pre_formatted = False
-        self.follows_fmt_tag = False
-        self.follows_trailing_space = False
+        self.font_size = pdf.font_size_pt
+        self.set_font(pdf.font_family or "times", size=self.font_size, set_default=True)
+        self._prev_font = (pdf.font_family, self.font_size, self.style)
+        self.pdf._push_local_stack()  # xpylint: disable=protected-access
+
+        self._pre_formatted = False  # preserve whitespace while True.
+        self._pre_started = (
+            False  # nothing written yet to <pre>, remove one initial nl.
+        )
+        self.follows_trailing_space = False  # The last write has ended with a space.
+        self.follows_heading = False  # We don't want extra space below a heading.
         self.href = ""
         self.align = ""
-        self.page_links = {}
         self.font_stack = []
         self.indent = 0
         self.bullet = []
-        self.font_size = pdf.font_size_pt
-        self.set_font(pdf.font_family or "times", size=self.font_size)
         self.font_color = tuple((255 * v for v in pdf.text_color.colors))
         self.heading_level = None
-        self.heading_sizes = dict(**DEFAULT_HEADING_SIZES)
         self.heading_above = 0.2  # extra space above heading, relative to font size
-        self.heading_below = 0.2  # extra space below heading, relative to font size
-        if heading_sizes:
-            self.heading_sizes.update(heading_sizes)
-        self.pre_code_font = pre_code_font
-        self.warn_on_tags_not_matching = warn_on_tags_not_matching
+        self.heading_below = 0.4  # extra space below heading, relative to font size
```

Comment:
Why doubling this value?

Reply:
I think this really visible in https://github.com/gmischler/fpdf2/blob/TextRegion/test/html/html_custom_heading_sizes.pdf

Reply:
This is notably visible on https://github.com/gmischler/fpdf2/blob/TextRegion/test/html/html_custom_heading_sizes.pdf

Reply:
Yes, as noted in the progress report, I was unable to exactly reproduce the line previous spacing in all cases (I also didn't always understand the logic behind it). Changing this factor brought the result reasonably close. Typically the layout will get slightly more compact, so the change won't cause additional page breaks.

Since changing the layout is kind of the point of the whole exercise, I'm not too woried about it...

Reply:
IMHO, after reviewing the differences between your branch and `master` regarding the reference PDF files in `test/html/`, the spacing has increased. Maybe it would still be worth to reduce this factor.

Reply:
Huh? In which files did you see increased spacing?
Did you make sure you are viewing both files at the same zoom level? Acrobat reader has the annoying habit of just changing that whenever it feels like it.

Reply:
* https://github.com/gmischler/fpdf2/blob/TextRegion/test/html/html_custom_heading_sizes.pdf VS https://github.com/gmischler/fpdf2/blob/master/test/html/html_custom_heading_sizes.pdf
* https://github.com/gmischler/fpdf2/blob/TextRegion/test/html/html_headings_line_height.pdf VS https://github.com/gmischler/fpdf2/blob/master/test/html/html_headings_line_height.pdf
* https://github.com/gmischler/fpdf2/blob/TextRegion/test/html/html_features.pdf VS https://github.com/gmischler/fpdf2/blob/master/test/html/html_features.pdf

Reply:
In all three examples, my version is slightly more compact than the one from master, when viewed at the same zoom level.
You really need to set the zoom explicitly, or the viewers (and browsers) will show you whatever random scale happens to fit into the window.